### ISSUE TITLE
Setup agent to support OLTP ingest does not work. 

### ISSUE THREAD

**Issue by samholder** [DataDog/datadog-agent#11664]
**Output of the info page (if this is a bug)**
```
Don't think this is a bug, just unsure how to configure correctly
```

**Describe what happened:**

This is running on Windows machine with agent installed (7.32.4.1) which reports in the logs

>Listening for traces at http://localhost:8126

and our application is sending traces via APM currently. 

I am trying to add support for open telemetry alongside the current APM stuff.

I followed the instructions here: https://docs.datadoghq.com/tracing/setup_overview/open_standards/#otlp-ingest-in-datadog-agent and added this config to the datadog.yml:

```
experimental:
  otlp:
    receiver:
      protocols:
        grpc:
        http:
```
When I restarted the agent I saw this in the logs:

>2022-04-13 15:10:48 BST | TRACE | WARN | (pkg/util/log/log.go:630 in func1) | Unknown key in config file: experimental.otlp.receiver.protocols.grpc
2022-04-13 15:10:48 BST | TRACE | WARN | (pkg/util/log/log.go:630 in func1) | Unknown key in config file: experimental.otlp.receiver.protocols.http

I searched little and found this issue: https://github.com/DataDog/helm-charts/issues/529

which seems to imply this this feature is no longer experimental, however I was not able to get this to function. Things I tried:

using this config in the agent datadog.yml:

```
otlp:
  receiver:
    protocols:
      grpc:
      http:
```

but same error as above basically.

Setting the following environment variables (based on other issue)

`OTEL_EXPORTER_OTLP_ENDPOINT` to `http://localhost:4317`
`DD_OTLP_HTTP_PORT` to 4317
`DD_OTLP_GRPC_PORT` to 4318
`OTLP_COLLECTOR` to `http://localhost:4317`

but after restarting the agent I see no other messages apart from the:

>Listening for traces at http://localhost:8126

which implies this did not work.
I also tried the app just in case.

The app is a .net 6 web app with this configration:

```
 services.AddOpenTelemetryTracing(
                builder =>
                {
                    builder
                        .SetSampler(new AlwaysOnSampler())
                        .AddSource("MySource")
                        .SetResourceBuilder(
                            ResourceBuilder.CreateDefault()
                                .AddService(serviceName: "MyService", serviceVersion: "1.0.0"))
                        .AddOtlpExporter(config =>
                        {
                            config.Endpoint = new Uri("http://localhost:4317");
                        })
                        .AddNServiceBusInstrumentation();
                });
```

**Describe what you expected:**

That following the instructions for enabling OTLP ingest would work correctly, or some alternative instructions are available if the feature is not experimental any more.

**Steps to reproduce the issue:**
As above

**Additional environment details (Operating System, Cloud provider, etc):**

locally hosted Windows Server 2019VM 
Datadog Agent 7.32.4.1
Datadog .NET tracer 64 bit 1.27.1

**Comment by samholder** [comment]
After a bit more investigation if I enable debug logging then I get this in the logs:

>2022-04-14 12:19:37 BST | TRACE | DEBUG | (pkg/trace/api/otlp.go:97 in Start) | OpenTelemetry gRPC receiver running on localhost:5003 (internal use only)

and if I configure things to point at this endpoint then I get some traces being logged, but they appear to get tagged as internal, so I assume this is not the right thing to be doing...

**Comment by guizmaii** [comment]
Same here :/

**Comment by pj-datadog** [comment]
Hi, I am a Product Manager at Datadog. We recently declared our OTLP Ingest in Datadog Agent (sending telemetry data from OTel SDK to DD Agent) as stable/GA, and is available with agent version 7.35. A very clear and descriptive documentation for this feature will be available in our public docs in the mid-next week. This should solve the setup problems you are facing but in case further support is needed, I will also be happy to connect you to our engineering team.
Thank you for choosing Datadog as your APM vendor of choice!

**Comment by guizmaii** [comment]
For anyone reading this issue, I figured my issue out.

The traces were correctly sent to Datadog. My configuration (explained here: https://github.com/DataDog/helm-charts/issues/529#issuecomment-1099421478) works perfectly well.

I wasn't finding my traces because I was looking for them in the `env: staging` in the APM UI of Datadog, while they were reported in the `env: none`.

To fix this `env` issue, I configured the `ResourceAttributes.DEPLOYMENT_ENVIRONMENT` in my `SdkTracerProvider` instance:

```scala
val serviceName = "my-service"
val env = "staging"
val resource = 
  Resource
    .builder()
    .put(ResourceAttributes.SERVICE_NAME, serviceName)
    .put(ResourceAttributes.DEPLOYMENT_ENVIRONMENT, environment)
    .build()

...    

val tracerProvider = 
  SdkTracerProvider
   .builder()
   .addSpanProcessor(spanProcessor)
   .setResource(resource)
   .build()    
```

**Comment by iamharvey** [comment]
> Hi, I am a Product Manager at Datadog. We recently declared our OTLP Ingest in Datadog Agent (sending telemetry data from OTel SDK to DD Agent) as stable/GA, and is available with agent version 7.35. A very clear and descriptive documentation for this feature will be available in our public docs in the mid-next week. This should solve the setup problems you are facing but in case further support is needed, I will also be happy to connect you to our engineering team. Thank you for choosing Datadog as your APM vendor of choice!

What is the difference between Datadog's otel agent and standard otel collector? Currently we are using otel-collector where the exporters are set to datadog. Will the release influences our current use?

**Comment by duxing** [comment]
hi folks, I've spent a good amount of time working on this topic recently and I'd like to share what I've found out here.

TL;DR; version of the issue: `agent` listens on `4317`/`4318`(`otlp` receiver) correctly but the traces were not sent to the `trace-agent` (on internal port `5003`)

my setup:
- "dd-agent": version `7.35.0`. `otlp` is enabled via environment variable: (tested with `datadog.yaml` as well, same result)
- `DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_GRPC_ENDPOINT=0.0.0.0:4317`
- `DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_GRPC_TRANSPORT: tcp`
- I've verified that datadog is listening on port `4317` with `netstat -apln | grep 4317`
- application: python flask + `opentelemetry-python` auto-instrumentation

before attempting using OTLP ingestion on `datadog-agent`, my application was auto-instrumented with `opentelemetry-exporter-datadog` (being deprecated): I was able to see APM traces on my datadog account

when I replaced `opentelemetry-exporter-datadog` with `opentelemetry-exporter-otlp` (pointing to the `4317` port), I lost APM traces on my datadog account. same API key was used compared to before and the rest of application code remains the same

some details for `opentelemetry-exporter-otlp` implementation:
- environment variable `OTEL_RESOURCE_ATTRIBUTES` is set for my application:
   - `OTEL_RESOURCE_ATTRIBUTES="service.name=my_app,deployment.environment=my_env,service.version=my_version"`
- compared to the implementation using `opentelemetry-exporter-datadog`:
   - `service.name` matches the value for `DD_SERVICE`
   - `deployment.environment` matches the value for `DD_ENV`
   - `service.version` matches the value for `DD_VERSION`
- `otlp_exporter = OTLPSpanExporter(endpoint="http://datadog:4317", insecure=True)` (`datadog` is the hostname in the `docker` network)
- span processor was configured (this is the same as using `opentelemetry-exporter-datadog`):
   - trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(otlp_exporter))
   - in addition to OTLP exporter, I also added a console exporter to debug
   - `trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))`


what I observed:
- telemetry data (traces, auto-instrumentation as well as manual) are showing correctly in console output.
- running `tcpdump` inside datadog agent container: `tcpdump -nnA -i any "port 4317" -s 0`: I was able to see the correct data in plain text coming to the `4317` port
- verified that `agent` is connected to `trace-agent` on port `5003`(internal grpc server for `agent` to export traces to `trace-agent` receiver)
- running `tcpdump` inside datadog agent container: `tcpdump -nnA -i any "port 5003" -s 0`: I don't see the telemetry data.

since the `agent` ingest `otlp` and export to `trace-agent` in `otlp`, I tried pointing my application otlp exporter to port 5003 instead and this failed too.

I've set `DD_LOG_LEVEL` to `trace` and inspected the `/var/log/datadog/agent.log` and `/var/log/datadog/trace-agent.log` but nothing looks suspicious. the only thing related is a "connection refused" error message when `agent` is trying to connect to `trace-agent` before `trace-agent` starts listening on the internal grpc port.

would be great to have a working example published from DataDog to help us understand what's missing for trace

**Comment by pj-datadog** [comment]
@iamharvey Today there are two methods a customer can use to send their telemetry data  to Datadog.

Method 1: [OTLP Ingest in Datadog Agent](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_the_agent/?tab=host)  - A way to send telemetry data from OTel SDKs directly to Datadog Agent, 

Method 2: [OTel Collector Datadog Exporter](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otel_collector_datadog_exporter/) - A way to send telemetry data from OTel SDKs to OTel Collector, which exports the data to Datadog Backend via a Datadog Exporter.

If you are using OTel Collector Datadog Exporter method, the release (GAing OTLP Ingest) will not influence your use.

NOTE: I am happy to announce that [OTLP Ingest in Datadog Agent](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_the_agent/?tab=host) is now GA/Stable with Datadog Agent version 7.35

**Comment by guizmaii** [comment]
> I am happy to announce that [OTLP Ingest in Datadog Agent](https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_the_agent/?tab=host) is now GA/Stable with Datadog Agent version 7.35

Not available yet in the Helm Chart ;)

**Comment by duxing** [comment]
@pj-datadog 
I love the idea of method 1 to make it easier to adopt. I've tested this setup but ran into an issue with traces (reported in #11737 ) and it seems like I'm not alone. 

I have a gut feeling that this is caused by misconfiguration instead of a bug in `datadog-agent`. would be great to have more examples / documentation to refer to

**Comment by pj-datadog** [comment]
@duxing Can you please reach out to [Datadog Support](https://help.datadoghq.com/hc/en-us) and open a zendesk ticket there. I asked my engineering team to look at your use-case, we feel we will need a debug flare. 

> would be great to have more examples / documentation to refer to
What kind of documentation do you have in mind? I can work with you to have that in our public documentation if you think that will be helpful for the larger community.
Thank You!

**Comment by duxing** [comment]
@pj-datadog 
a support ticket has been open since a few days ago:
https://help.datadoghq.com/hc/en-us/requests/789265?page=1

details are cross-referenced

> we feel we will need a debug flare
I don't think you need that from me. see README from the [repo](https://github.com/duxing/datadog-otlp) I linked in the referred issue #11737 
once you `git clone` the repo, you should be able to get what you need.

> What kind of documentation do you have in mind?
mainly examples I guess. the only example I found was the gist that @gbbr provides (for golang). more examples supporting more languages would be really nice to have.

**Comment by mackjmr** [comment]
@guizmaii currently you can set this manually via the environment variables in helm (https://docs.datadoghq.com/tracing/setup_overview/open_standards/otlp_ingest_in_the_agent/?tab=kuberneteshelm), but we are actively working on a dedicated configuration section to make this easier

**Comment by tohadar** [comment]
I tried to use method2 with the section of '[Alongside Datadog Agent](https://docs.datadoghq.com/tracing/trace_collection/open_standards/otel_collector_datadog_exporter/#alongside-the-datadog-agent)' and it didn't work.
Tried to deploy opentelemetry as Damonset alongside Datadog agent Damonset but wasn't able to succeed since both Damonset try to listen on the same port of the hosts(4317,4318).
Also tried to deploy opentelemtry as deployment and to use the otlp exporter:

exporters:
  otlp:
    endpoint: "${HOST_IP}:4317"

But got tls errors.
Did anyone was able to configure it and send traces?

**Comment by mx-psi** [comment]
Note that you don't need to deploy the OpenTelemetry Collector to use OpenTelemetry. You can just use the Datadog Agent if you want and point your application to send telemetry data to the Agent.

> Also tried to deploy opentelemtry as deployment and to use the otlp exporter:
> But got tls errors.

If you still want to use the Collector and the Agent, you can disable TLS by doing this:

```
exporters:
  otlp:
    endpoint: "${HOST_IP}:4317"
    tls:
      insecure: true
```

This should be safe to do if communication happens locally.

**Comment by tohadar** [comment]
> Note that you don't need to deploy the OpenTelemetry Collector to use OpenTelemetry. You can just use the Datadog Agent if you want and point your application to send telemetry data to the Agent.
> 
> > Also tried to deploy opentelemtry as deployment and to use the otlp exporter:
> > But got tls errors.
> 
> If you still want to use the Collector and the Agent, you can disable TLS by doing this:
> 
> ```
> exporters:
>   otlp:
>     endpoint: "${HOST_IP}:4317"
>     tls:
>       insecure: true
> ```
> 
> This should be safe to do if communication happens locally.

tls.insecure: true 
in opentelemetry helm chart?

**Comment by vivere-dally** [comment]
Can the agent also ingest metrics?

**Comment by TimoSchmechel** [comment]
@duxing did you deploy the DD agent via helm? 

I am currently facing the same issue, but only after changing over to the Datadog Operator.
My previous setup was, DD agent via helm which received otlp from the OTEL collector. All I did was change the DD agent to be deployed via the DatadogAgent resource from the operator and now the agent no longer forwards traces that it receives from the collector. 

I've "fixed" the issue by just exporting to DD directly by using the DD exporter inside the OTEL collector.

**Comment by duxing** [comment]
@TimoSchmechel 
not for my set up.
I'm running my setup locally with `docker-compose` and deploying the same change through `helm` to staging/production environment would have been the next step.

I think the issue is in the agent image/binary itself, not the chart. however, there's not a good way to reproduce this issue. the test project (setup with `docker`) I setup consistently surface this issue on my end, but did not yield any error on datadog side.

**Comment by maxwoods** [comment]
In my case the issue was the trace-agent (component that listens on 5003/tcp) is disabled by default when using the Datadog operator. Traces were making it to the external otlp endpoints (4317/4318) on the agent, then getting blackholed at the trace-agent port (5003).

Setting `features.apm.enabled` to `true` turned on the trace-agent container and fixed my issue.

@pj-datadog this may be helpful to document here for those of us that didn't already have APM enabled: https://docs.datadoghq.com/opentelemetry/otlp_ingest_in_the_agent/?tab=host

**Comment by changhyuni** [comment]
Has this issue been resolved?
I followed your example for helm chart but it doesn't work
```yaml
      datadog:
        otlp:
          receiver:
            protocols:
              http:
                enabled: true
                endpoint: "0.0.0.0:4318"
                useHostPort: true
```
I can see only this log
```
UTC | TRACE | DEBUG | (pkg/trace/api/otlp.go:91 in Start) | Listening to core Agent for OTLP traces on internal gRPC port (http://0.0.0.0:5003, internal use only). Check core Agent logs for information on the OTLP ingest status.
```
Looking at the pod manifest, it looks like you have the port specification and variables set up well.

**Comment by ansraliant** [comment]
I'm running agent version `7.60` and currently experiencing this issue just as @duxing explained.

The traces from the application are being send correctly to the datadog-agent, but, they are not being sent to the `trace-agent`.

The OTLP receiver is listening at port `4317`, and the trace-agent is listening at port `5003`

Using the OTEL collector with the datadog-exporter, did solve the issue, at least for now.

**Comment by half0wl** [comment]
I'm unable to get OTEL collection working as well.

The only thing that stands out are these logs on failing to connect to `:5001` (I believe that's the trace agent?):
```
2025-03-13 11:31:16 UTC | TRACE | INFO | (pkg/config/remote/client/client.go:400 in pollLoop) | retrying the first update of remote-config state (rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:5001: connect: connection refused")
2025-03-13 11:31:16 UTC | SYS-PROBE | INFO | (pkg/config/remote/client/client.go:400 in pollLoop) | retrying the first update of remote-config state (rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp [::1]:5001: connect: connection refused")
2025-03-13 11:31:16 UTC | SYS-PROBE | INFO | (comp/core/tagger/impl-remote/remote.go:586 in func1) | unable to establish stream, will possibly retry: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp :5001: connect: connection refused"
2025-03-13 11:31:16 UTC | SYS-PROBE | INFO | (comp/core/workloadmeta/collectors/internal/remote/generic.go:171 in func1) | unable to establish stream, will possibly retry: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp :5001: connect: connection refused"
2025-03-13 11:31:17 UTC | SYS-PROBE | INFO | (comp/core/tagger/impl-remote/remote.go:586 in func1) | unable to establish stream, will possibly retry: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp :5001: connect: connection refused"
2025-03-13 11:31:17 UTC | TRACE | INFO | (backoff@v2.2.1+incompatible/retry.go:37 in RetryNotify) | unable to establish stream, will possibly retry: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp :5001: connect: connection refused"
2025-03-13 11:31:17 UTC | SYS-PROBE | INFO | (comp/core/workloadmeta/collectors/internal/remote/generic.go:171 in func1) | unable to establish stream, will possibly retry: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp :5001: connect: connection refused"
```

Traces (via Node dd-trace) works.

I tried firing off a test trace to the OTLP receiver and it seemed to go through, but I don't see it anywhere in Datadog:
```
root@a71d866b1162:/app# curl -X POST http://hs-dd-agent.railway.internal:4318/v1/traces \
  -H "Content-Type: application/json" \
  -d '{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": { "stringValue": "test-service" }
          },
          {
            "key": "deployment.environment",
            "value": { "stringValue": "test" }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "manual-test"
          },
          "spans": [
            {
              "traceId": "8aecfcbd3cdb284fb1bcd381a532c118",
              "spanId": "5759e988c1fd2ea3",
              "name": "test-span",
              "kind": 1,
}'] } ] } ] } } "code": 1: { "stringValue": "/api/test" }
{"partialSuccess":{}}
```

I'm running the Agent on Docker using the `datadog/agent:7.63.3` image with the following configuration:

```
DD_API_KEY="..."
DD_APM_COMPUTE_STATS_BY_SPAN_KIND="true"
DD_APM_PEER_TAGS_AGGREGATION="true"
DD_LOGS_ENABLED="true"
DD_OTLP_CONFIG_LOGS_ENABLED="true"
DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_HTTP_ENDPOINT="[::]:4318"
DD_OTLP_CONFIG_RECEIVER_PROTOCOLS_GRPC_ENDPOINT="[::]:4317"
DD_APM_ENABLED="true"
DD_APM_NON_LOCAL_TRAFFIC="true"
```

**Comment by mertcankloian** [comment]
Hi @half0wl , your configuration for both sides ( application otelsdk, datadog otlp) looks correct but somehow you got tcp Error on 5001 which is IPc port of DD. (internal Process Communication) . I believe you have some issue on dd side and that is not directly related with otlp. Did you check the network for 5002 port. It should open as ınboud or try to change that port for testing. 

Here docs. 
https://docs.datadoghq.com/agent/configuration/network/#inbound

**Comment by pierreprinetti** [comment]
Tentatively closing. Please reopen if the issue is still valid.
